### PR TITLE
Add input/output schemas to runnables

### PR DESCRIPTION
**Pull request by nfcampos**
This adds `input_schema` and `output_schema` properties to all runnables, which are Pydantic models for the input and output types respectively. These are inferred from the structure of the Runnable as much as possible, the only manual typing needed is 
- optionally add type hints to lambdas (which get translated to input/output schemas)
- optionally add type hint to RunnablePassthrough

These schemas can then be used to create JSON Schema descriptions of input and output types, see the tests

- [x] Ensure no InputType and OutputType in our classes use abstract base classes (replace with union of subclasses)
- [x] Implement in BaseChain and LLMChain
- [x] Implement in RunnableBranch
- [x] Implement in RunnableBinding, RunnableMap, RunnablePassthrough, RunnableEach, RunnableRouter
- [x] Implement in LLM, Prompt, Chat Model, Output Parser, Retriever
- [x] Implement in RunnableLambda from function signature
- [x] Implement in Tool

### CODE REVIEW

**Review by eyurtsev — APPROVED** [review]
This looks great!

File: libs/langchain/langchain/schema/runnable/base.py [hunk 1]
```diff
@@ -67,6 +69,42 @@ class Runnable(Generic[Input, Output], ABC):
     """A Runnable is a unit of work that can be invoked, batched, streamed, or
     transformed."""
 
+    @property
+    def InputType(self) -> Type[Input]:
+        for cls in self.__class__.__orig_bases__:  # type: ignore[attr-defined]
```
Comment (eyurtsev) [inline libs/langchain/langchain/schema/runnable/base.py:74]: Is there an example of a runnable where the default implementation returns something? Most of the runnables seem to need to overload the input type

Comment (nfcampos) [reply libs/langchain/langchain/schema/runnable/base.py:75]: see eg base Retriever

File: libs/langchain/langchain/schema/messages.py [hunk 2]
```diff
@@ -277,6 +278,9 @@ def __add__(self, other: Any) -> BaseMessageChunk:  # type: ignore
         return super().__add__(other)
 
 
+AnyMessage = Union[AIMessage, HumanMessage, ChatMessage, SystemMessage, FunctionMessage]
```
Comment (eyurtsev) [inline libs/langchain/langchain/schema/messages.py:281]: Nice

**Review by eyurtsev — APPROVED** [review]

**Pull request merged**